File: example/chsh.jl
using SPDC
using Optim

function chsh(p::Vector{Float64};N=1,ηA=1.0,ηB=1.0)
	X = 2 # Number of Alice's input
	Y = 2 # Number of Bob's input
	p = param(p,X,Y;N=N,ηA=ηA,ηB=ηB)
	AB = spdc_correlators(p,X,Y)
	score = AB[1,1]+AB[1,2]+AB[2,1]-AB[2,2]
	return score
end

function optimize_chsh(;x0=rand(10),N=1,η=1.0)
	r = optimize(x -> -chsh(x,N=N,ηA=η,ηB=η), x0, Optim.Options(iterations=2000))
	return r
end

function chsh_with_loss()
	scores = []
    η = 1.0
    r = optimize_chsh()
    while -r.minimum ≥ 2.0
        push!(scores,[η,-r.minimum])
		@info scores[end]
        η -= .01
        r = optimize_chsh(η=η,x0=r.minimizer)
	end
	return scores
end

File: example/chsh.py
from julia import SPDC
import numpy as np
from scipy.optimize import minimize

def chsh(p,η=1.0):
    """ Compute the CHSH value for parameter p

    Param
    -----
        p : np.array 
            Circuit parameters, 1-d float array.
        η : float
            Photondetector efficiency

    Return
    ------
       chsh_score : float
            The computed CHSH score
    """
    X = 2 # Number of measurement of Alice
    Y = 2 # Number of measurement of Bob
    # Translate the parameter to a NamedTuple
    p = SPDC.param(p,X,Y,ηA=η,ηB=η)
    # Compute all the correlators <A_xB_y>
    AB = SPDC.spdc_correlators(p,X,Y)
    # Compute the CHSH score
    chsh_score = 0.0
    for x in range(2):
        for y in range(2):
            chsh_score += (-1)**(x*y)*AB[x,y] 
    return chsh_score

def optimize_chsh(η=1.0,x0=np.random.random(10)):
    """ Optimize the CHSH score. """
    res = minimize(lambda x: -chsh(x,η=η),
            x0,
            method="BFGS")
    return res

def chsh_with_loss(step=1e-2):
    """ Optimized CHSH score for decreasing efficiency value """
    scores = []
    η = 1.0
    r = optimize_chsh()
    while -r.fun >= 2.0:
        scores.append([η,r.fun])
        η -= step
        r = optimize_chsh(η=η,x0=r.x)
        print(scores[-1])
    return np.array(scores)
